Cache UPE intents for reuse

anu-rock opened the pull request:
Fixes #2703
Fixes #3602

#### Changes proposed in this Pull Request

Mounting the UPE element requires us to create a payment or setup intent in advance in the front-end code. This happens on the Checkout page. When a non-UPE payment method is used or the page is reloaded, the previously created intent is wasted and a new one is created. This has the potential of resulting in a large number of abandoned intents even for just one merchant.

This PR fixes this behavior by caching the newly created intent in a cookie. Once cached, it is then reused in the next mounting of UPE. The flow looks like this (driven by JS code):

![image](https://user-images.githubusercontent.com/1288616/150512260-504e6646-aedf-4bde-9576-f59907dc69f6.png)

Furthermore:

* Since a payment intent is dependent on the cart's total amount, the logic ties an intent with cart hash.
* The cookies expire ~~30~~ 10 mins after creation.

#### Testing instructions

Testing will require you to enable UPE first. Do this by using the "Enable the new Stripe checkout experience" option on the Settings page. Update the store currency as per the payment methods you want to test. Eg. Giropay, iDEAL, etc. will require Euro as the store currency.

##### Payment Intent

* Create a new order in the front-end.
* Open Network tab in browser's dev tools, and visit the Checkout page.
* On this page, notice an XHR call to `create_payment_intent`.
* Also confirm that a new cookie `wcpay_upe_payment_intent` is created.
* Reload the page. No fresh XHR call to create_payment_intent should be observed.
* Complete the payment. The above cookie should be now removed.
* Also check if the cookie is removed when a payment fails, for example when:
  * using a [credit card that is immediately declined](https://stripe.com/docs/testing#cards-responses)
  * using a [card that requires further authentication](https://stripe.com/docs/testing#regulatory-cards) and eventually the transaction fails
  * using a redirection payment method (e.g. Giropay, SEPA, iDEAL, Sofort, etc.) and eventually the transaction fails
  * using a saved payment method that fails

**FYI:** The intent id retrieved from cache or received as a response from `create_payment_intent` is used only when a new payment method is used. Using saved cards will create a new intent on the server, resulting in the UPE intent being abandoned.

##### Setup Intent

* Open Network tab in browser's dev tools, and visit My account → Payment methods → Add payment method.
* On this page, notice an XHR call to `init_setup_intent`.
* Also confirm that a new cookie `wcpay_upe_setup_intent` is created.
* Reload the page. No fresh XHR call to create_payment_intent should be observed.
* Add a new payment method. The above cookie should be now removed.

#### Scenarios I have testing

* New card - successful payment
* New card - declined
* New card - authentication failed
* Saved card - successful payment
* Saved card - authentication failed
* New SEPA - successful payment
* Saved SEPA - successful payment
* Giropay - successful payment
* Giropay - failed payment

-------------------

- [x] Added changelog entries to both `readme.txt` and `changelog.txt`
- [x] Covered with tests
- [ ] Tested on mobile (does not apply)

**Post merge**

- [x] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) : https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions-for-WC-Payments-3.7.0#cache-upe-intents-for-reuse

anu-rock committed:
Cache UPE intents in cookies

FangedParakeet commented on client/checkout/classic/upe.js:
```diff
@@ -7,12 +7,16 @@ import './style.scss';
 import enqueueFraudScripts from 'fraud-scripts';
 import { getFontRulesFromPage, getAppearance } from '../upe-styles';
 import { getTerms } from '../utils/upe';
+import { getCookieValue } from '../../utils';
```
Pedantic peeve, but you could have put this function inside `../utils/upe.js`.

anu-rock replied:
Hehe, yes!

anu-rock replied:
On another thought, `getCookieValue` is a generic utility not limited to UPE. Perhaps code outside of UPE might benefit from this generic function at some point?

FangedParakeet replied:
On its own `getCookieValue` is a pretty generic looking utility, although, that being said, I don't think it has any purpose outside of the UPE. At least, I don't think there is a reason to maintain a static session for payment intents without the UPE, since the intent will regularly never exist until the `Pay for order` button is selected.

I did originally admit that this was a pedantic complaint and honestly the only reason I made it was because it took me an extra five seconds to locate this function. 😅 I'm not too bothered where it ends up, so I'll leave its final resting place up to your own discretion.

anu-rock replied:
Thanks :) I'll leave it in the common utils for now so if anyone is looking to access any cookie they have an easier way than parsing `document.cookie`.

FangedParakeet reviewed (changes requested):
Nice one: I like the refactor on `mountUPEElement` and the general gist of what's going on here! This generally worked!...But I also managed to break it in a couple ways as well. 😁

Here are places where saving the payment intent is currently not supported, which I'd be happy to solve in other issues:
 - Blocks checkout with UPE
 - Order Pay checkout with UPE
 - Add/Change payment method from My Accounts with UPE

I was a little surprised that the last one didn't appear to be working, as I could see the setup intent cookie was being stored, but whenever I refreshed the page, a new intent was still created.

However, there were a couple oddities that I managed to engineer from the regular checkout that hopefully you'll be able to resolve.

**Setup intents at checkout**
To force a setup intent to be used during a regular checkout, you need to checkout for a subscription product with a free-trial: for this checkout, we will need to collect payment information, but not charge it immediately--thus necessitating the setup intent.

When I checked out with this item, refreshing the page still refreshed the intent. This could be because the cookie was being stored as a payment intent and retrieved as a setup intent--or perhaps for other reasons. I explored this in another comment, so hopefully you'll be able to figure it out!

**Failing transactions**
If you force a transaction to fail and then return to the checkout, something a little interesting happens: the UPE loads, but it only shows inputs for the previous payment method used and it is now impossible to change this!

Here are a few ways you can make this happen:
 - Checkout using a [credit card that is immediately declined](https://stripe.com/docs/testing#cards-responses)
 - Checkout with a [card that requires further authentication](https://stripe.com/docs/testing#regulatory-cards) and manually fail the transaction
 - Checkout with a redirection payment method (e.g. Giropay, SEPA, iDEAL, Sofort, etc.) and manually fail the transaction after the redirect.
 
 When you return the checkout and try reload the UPE, it will only load the failed payment method. I _think_ this is because prior to confirming the intent, [we update the intent to only contain the selected payment method type](https://github.com/Automattic/woocommerce-payments/blob/develop/includes/payment-methods/class-upe-payment-gateway.php#L197), instead of the array of all enabled payment methods. If you try to reload the UPE with this intent, which now only has a single `payment_method_type`, the UPE will only support that payment method, ignoring the other enabled payment methods. 
 
 I think one solution to this would be to move the `remove_upe_intent_cookies` function to a few more places, instead of relying on the `woocommerce_order_payment_status_changed` hook. There are a few points where an intent may no longer be reusable: off the top of my head, this could happen inside `UPE_Payment_Gateway::process_redirect_payment`, inside one of the Stripe webhooks in the `WC_REST_Payments_Webhook_Controller` class, or maybe inside `WC_Payment_Gateway_WCPay::process_payment_for_order` if a saved payment method causes an error,  or inside  `WC_Payment_Gateway_WCPay::update_order_status_from_intent` if a payment method requiring authentication causes an error.
 
I've dropped a lot of information here and I don't expect you to be familiar with all of it, so please let me know if I can help you understand any of the above any better or if you have any thoughts about the things I brought up. Cheers!

anu-rock commented on includes/payment-methods/class-upe-payment-gateway.php:
```diff
@@ -46,6 +46,10 @@ class UPE_Payment_Gateway extends WC_Payment_Gateway_WCPay {
 
 	const UPE_APPEARANCE_TRANSIENT = 'wcpay_upe_appearance';
 
+	const COOKIE_UPE_PAYMENT_INTENT = 'wcpay_upe_payment_intent';
+
+	const COOKIE_UPE_SETUP_INTENT = 'wcpay_upe_setup_intent';
```
Indeed, that sounds like a good idea!

anu-rock replied:
Done ✅

anu-rock commented on includes/payment-methods/class-upe-payment-gateway.php:
```diff
@@ -1068,4 +1086,45 @@ public function log_payment_error_ajax() {
+	 * @param string $intent_id     The setup intent id.
+	 * @param string $client_secret The setup intent client secret.
+	 */
+	private function create_upe_setup_intent_cookie( string $intent_id = '', string $client_secret = '' ) {
+		$cookie_val = $intent_id . '-' . $client_secret;
+		wc_setcookie( self::COOKIE_UPE_SETUP_INTENT, $cookie_val, time() + MINUTE_IN_SECONDS * 30 );
+	}
```
My uneducated assumption was that there was a way to trigger the request to create a setup intent from the Checkout page. Based on your comments, it looks like this only happens on My Accounts pages or when checking out a subscription product with a free trial. There's certainly a lot to learn about the various possibilities and flows 😅 

At any rate, I wanted to keep different cookies for payment and setup intents because one depends on the cart's current state while the other does not. Also, it's a bit cleaner. Hence, two functions for two cookies.

anu-rock replied:
If you found my explanation reasonable, are we good to mark this conversation resolved?

FangedParakeet replied:
Cool! I don't really have a preference here, as long as it works. 😄 Keeping different cookies does sound reasonable enough, for the reasons you've shared. Resolving contentedly.

anu-rock commented:
@FangedParakeet Thanks a ton for trying out my changes and the detailed feedback. It's very comprehensive, and many things are certainly new to me 😉 I'll take some time to experiment more based on your suggestions and come back once the changes are ready for a final review.

anu-rock committed:
Keep cookie constants at one place

Using the getConfig utility, PHP constants can be accessed in JS.

anu-rock committed:
Merge branch 'develop' into fix/3602-redundant-intent-checkout-page

anu-rock commented:
@FangedParakeet I tested the setup intent flow today on the `/my-account/add-payment-method/` page.

> When I checked out with this item, refreshing the page still refreshed the intent.

It worked fine for me. I could verify that when the cookie was present, no network call was sent to create a new setup intent. I checked in FF and Chrome. Could you give it another shot?

Also, any tips on picking the relevant hooks to delete the setup intent cookie? 😄

anu-rock committed:
Fix payment intent cookie with setup intent value

This happens when checking out a subscription product that has a free trial.

anu-rock commented:
@FangedParakeet 

> To force a setup intent to be used during a regular checkout, you need to checkout for a subscription product with a free-trial: for this checkout, we will need to collect payment information, but not charge it immediately--thus necessitating the setup intent.

> When I checked out with this item, refreshing the page still refreshed the intent. This could be because the cookie was being stored as a payment intent and retrieved as a setup intent--or perhaps for other reasons.

You are right. In this case, the `UPE_Payment_Gateway::create_payment_intent` method returns a setup intent but the payment intent cookie gets set with setup intent id as its value. But for me refreshing the page was not refreshing the intent because of the cookie.

At any rate, to avoid this confusing mismatch, I'm now creating the payment intent cookie only when the response starts with "pi." That means, refreshing the page in this case will now always refresh the intent. IMO, the subscription-with-free-trial product is not as common as standard product types. So, it's okay to let this scenario suffer for now.

FangedParakeet commented:
> It worked fine for me. I could verify that when the cookie was present, no network call was sent to create a new setup intent. I checked in FF and Chrome. Could you give it another shot?

Verified! This is working properly for me now. In fact, I have a feeling that I might have sabotaged your code myself somewhere, while testing things, and caused this to fail and myself to complain in the first place. 🤦

FangedParakeet commented:
> Also, any tips on picking the relevant hooks to delete the setup intent cookie? 😄

Haha, off the top of my head...no. But I suppose adding your own hook certainly could be an option. I _think_ all that needs to happen is that if the transaction fails, when we return to the checkout we need to have an intent with `payment_method_types` on the intent set to an array of all available/enabled payment methods. Either creating a new payment intent or updating the existing payment intent would accomplish this. There are pros and cons to either approach, so I'll leave it up to you to explore that. 

In terms of where that action needs to take place, I think there are two cases where this will happen: inside the payment gateway where there is an error and potentially inside the webhooks, should they fire fast enough (which I believe is technically possible). I think I mentioned the guilty functions somewhere above, but let me know if you need more help tracking down these trapdoors and I'll dig a little further to give you more accurate specifics!

anu-rock committed:
Remove cookies more effectively

Using unset along with setcookie increases the likelihood of clean
removal of cookies.

This also fixes the issue where cookies weren't deleted on using a new
payment method on the Checkout page.

anu-rock committed:
Decrease cookie expiry time to 10 mins

This minimizes chances of a stale cookie, that could not be successfully
removed, from causing nasty errors.

anu-rock committed:
Remove UPE cookies on adding a new payment method

The hook woocommerce_after_account_payment_methods is only used as an
approximation. It is triggered on load of the Payment Methods page in
My Account. This is the page where user is redirected after a new payment
method is successfully added.

anu-rock committed:
Split remove cookies method into separate methods

anu-rock committed:
Remove cookie on failure of redirection payment

Giropay, SEPA, Sofort, iDEAL, etc.

anu-rock committed:
Remove cookie on failure of a saved payment method

anu-rock committed:
Fix the UPE gateway import

anu-rock committed:
Remove cookie on failures captured on client-side

Client-side = JS

Examples of such errors include:
* a new card that is declined
* a new card that requires authentication and eventually fails it

Such failures are not captured on server-side (PHP code) because these
scenarios are directly handled by Stripe's JS library.

anu-rock commented on includes/payment-methods/class-upe-payment-gateway.php:
```diff
@@ -1095,8 +1095,12 @@ public function log_payment_error_ajax() {
 			}
 			$order->add_order_note( $order_note );
 
+			self::remove_upe_payment_intent_cookie();
```
This is the only place I found to handle "card declined" and "authentication failed" scenarios.

cc @FangedParakeet

anu-rock replied:
At least in my debugging, the control never went to `WC_Payment_Gateway_WCPay::update_order_status` on client-side errors (card declined or auth failed). I just re-checked to be sure. Still nothing :(

FangedParakeet reviewed (approved):
Left a few comments, but I re-tested the pesky edge-cases from before and everything appeared to be working as expected this time! Cheers for spending so much time on this in response to my initial lengthy feedback, I think this is a pretty nice ingredient to add to the intoxicating concoction that is the UPE. Nice one.

anu-rock committed:
Use wc_setcookie for removing UPE cookies

anu-rock commented on includes/payment-methods/class-upe-payment-gateway.php:
```diff
@@ -1074,4 +1110,57 @@ public function log_payment_error_ajax() {
+	public static function remove_upe_payment_intent_cookie() {
+		if ( isset( $_COOKIE[ self::COOKIE_UPE_PAYMENT_INTENT ] ) ) {
+			unset( $_COOKIE[ self::COOKIE_UPE_PAYMENT_INTENT ] );
+			setcookie( self::COOKIE_UPE_PAYMENT_INTENT, '', time() - HOUR_IN_SECONDS, '/' );
```
No reason. I guess I incorrectly assumed that `setcookie` would work better for removing cookies since it accepted the path param which I read at a few places was more effective in removing cookies.

Fixed in a799644

kalessil commented on includes/payment-methods/class-upe-payment-gateway.php:
```diff
@@ -46,6 +46,14 @@ class UPE_Payment_Gateway extends WC_Payment_Gateway_WCPay {
+
+	const COOKIE_UPE_SETUP_INTENT = 'wcpay_upe_setup_intent';
+
+	const COOKIE_UPE_EXPIRES_IN_MINS = 10;
```
10 mins only, any concerns increasing this value? Perhaps a couple of hours?

anu-rock replied:
My concern is noted in 7c20418. In a nutshell, an over-lived cookie may cause undesired side effects, [like this](https://github.com/Automattic/woocommerce-payments/pull/3684#pullrequestreview-858935708):

>  If you force a transaction to fail and then return to the checkout, something a little interesting happens: the UPE loads, but it only shows inputs for the previous payment method used and it is now impossible to change this!

I have tried to cover as many failure scenarios as I could think of, with help from @FangedParakeet of course. But in case an edge case got missed, a lower TTL would help mitigate side effects.

anu-rock committed:
Merge branch 'develop' into fix/3602-redundant-intent-checkout-page

kalessil reviewed (commented):
As much as I could learn from this PR:
- it is great that cleaning up the cookies is concerned
- we need to add change logs entries
- we'll need to communicate re-testing UPE before release

anu-rock commented on includes/payment-methods/class-upe-payment-gateway.php:
```diff
@@ -91,6 +99,9 @@ public function __construct( WC_Payments_API_Client $payments_api_client, WC_Pay
 		add_action( 'wp', [ $this, 'maybe_process_upe_redirect' ] );
 
+		add_action( 'woocommerce_order_payment_status_changed', [ 'WCPay\Payment_Methods\UPE_Payment_Gateway', 'remove_upe_payment_intent_cookie' ], 10, 0 );
+		add_action( 'woocommerce_after_account_payment_methods', [ $this, 'remove_upe_setup_intent_cookie' ], 10, 0 );
```
The removal of setup intent cookie on successfully adding a new payment method on My account → Payment methods page. This is described in ed43129.

kalessil replied:
I see, thanks for clarifying!

anu-rock committed:
Make payment intent id prefix more robust

anu-rock commented on includes/payment-methods/class-upe-payment-gateway.php:
```diff
@@ -222,7 +233,15 @@ public function create_payment_intent_ajax() {
+			$response = $this->create_payment_intent( $order_id );
+
+			// Cache the intent id in cookie.
+			if ( strpos( $response['id'], 'pi' ) === 0 ) { // response is a payment intent (could possibly be a setup intent).
```
Indeed, that's cleaner. Fixed in c99be6b.

kalessil reviewed (approved):
LGTM! I did only code review though, if more manual testing is expected, let me know, please.

anu-rock commented:
Thanks, @kalessil :) I think we're good for now as @FangedParakeet and I have done lots of manual testing. But if someone is up for testing, I've mentioned a few scenarios in the PR description.

anu-rock committed:
Merge branch 'develop' into fix/3602-redundant-intent-checkout-page

anu-rock committed:
Update changelog
